### PR TITLE
fix(storage): tolerate malformed legacy session headers during import

### PR DESCRIPTION
**Pull request by sunheyi6**
## Summary

A malformed legacy `session.jsonl` header no longer blocks startup migration or hides every valid session. The importer skips only repairable legacy-content failures, writes no deletion tombstone, and retries the session on a later launch so a repaired header can be imported.

The error boundary now preserves the existing fail-closed behavior for filesystem failures and invalid current-format `session_transcript` markers. A private structured error identifies only legacy framing, JSON, and header-decoding failures; message matching, cause-chain inspection, and filesystem error-code heuristics have been removed.

The public `SessionStore` tests cover empty and truncated files, a complete valid header beyond the 1 MiB limit, the reported `turn_state` first-record failure, invalid current marker identities/schema versions, repair and reopen, no tombstones, and no partial import on environmental or marker failures. Repeated recovery cases are table-driven.

## Verification

- `npm --workspace @maka/storage test` — 478 passed, 1 existing Windows-only test skipped
- `npm run format:check` — passed
- `npm run lint` — passed
- `npm --workspace @maka/storage run typecheck` — passed
- Targeted `session-metadata-transfer` and `sqlite-session-store` suites — 18 passed
- Mutation check: temporarily raising the production limit to 2 MiB made the >1 MiB regression test fail by importing the oversized valid header, confirming the test isolates the size cap; the production limit was restored before commit
- GitHub CI on `d1fae87f7` — typecheck, test, and e2e passed

## Root cause

The legacy migration originally treated every non-`ENOENT` read/decode failure as fatal. An intermediate fix then wrapped all non-filesystem failures as malformed legacy headers, which also swallowed validation errors from current-format transcript markers. The final implementation assigns recoverable status at the source: legacy framing/JSON/header errors use one private error type, while current marker validation and real I/O errors propagate unchanged.

### CODE REVIEW

**Review by Astro-Han — APPROVED** [review]
The malformed-header recovery works as intended. I found two non-blocking issues:

- [P2] `readLegacySessionMetadataEntry()` wraps every failure from opening, reading, parsing, decoding, and closing the file as `Invalid legacy session header`. `isMalformedLegacySessionHeader()` then accepts that outer message by itself, so a transient `EACCES`, `EIO`, or close failure is treated as corrupt data and permanently tombstones the session. Classify the underlying parse or validation errors instead of the generic wrapper. A test that injects a non-`ENOENT` filesystem failure would lock this down.
- [P2] Malformed-session tombstones are written through `destination.remove()` while the scan is still in progress, before `importEntries()` starts its transaction. If a later directory has a missing transcript marker or the final import conflicts, the import rejects but the earlier tombstone remains committed. A repaired legacy header will then stay suppressed on the next startup. Commit the tombstones with the successful import, or roll them back when the scan or import fails. The current test covers only the success path.

Approving because neither issue is a P0 or P1, but both cases should be covered before this migration path is treated as failure-safe.

File: packages/storage/src/session-metadata-transfer.ts [hunk 1]
```diff
@@ -106,6 +121,27 @@ function isNotFound(error: unknown): boolean {
...
+ * `"Invalid legacy session header at <path>"` with the original error as
+ * `cause`.  We walk the cause chain looking for the signature messages
+ * produced by `decodeSessionHeader` / `normalizeSessionHeader`.
+ */
+function isMalformedLegacySessionHeader(error: unknown): boolean {
+  let current = error;
+  while (current && typeof current === 'object') {
+    if (current instanceof Error) {
+      const msg = current.message;
+      if (msg.startsWith('Invalid legacy session header at ')) return true;
```
Comment (Astro-Han) [inline packages/storage/src/session-metadata-transfer.ts]: [P1] This outer message wraps every error from `readLegacySessionMetadataEntry()`, including open, read, and close failures. A transient `EACCES`, `EIO`, or `EMFILE` error therefore reaches the malformed-header branch and permanently tombstones a valid session.

Only JSON and legacy-header decode failures should be recoverable here. Filesystem errors and invalid current-format markers should still propagate. Add a non-`ENOENT` I/O test for this boundary.

**Review by Astro-Han — COMMENTED** [review]
Skipping a bad legacy header is the right behavior, but this patch can turn a read failure into permanent deletion state.

[P2] `destination.remove()` writes the same tombstone used for user deletion. Once that happens, `importEntries()` will keep skipping the session even if its header is repaired. For malformed legacy content, skip the session without calling `remove()`. If persistent quarantine is required, it needs separate semantics from deletion.

[P2] `remove()` runs during the scan, before the final `importEntries()` transaction. If a later marker check or import fails, the earlier tombstone remains committed. That leaves a partial migration even though the function promises a single transaction. Keep the scan free of writes, or commit quarantine state with the import batch.

**Comment by sunheyi6** [comment]
Thanks for the review! I've pushed a commit addressing both P2 issues:

## P2-1: Classify underlying errors instead of generic wrapper

`readLegacySessionMetadataEntry` now calls `readFirstJsonlRecord` outside the try-catch, so filesystem I/O errors (EACCES, EIO, EISDIR, etc.) propagate without being wrapped as `"Invalid legacy session header"`. Only parse/decode failures (JSON.parse, decodeSessionHeader, decodeSessionTranscriptMarker) are wrapped, so `isMalformedLegacySessionHeader()` no longer matches transient filesystem errors.

**Test added**: `does not tombstone a session when a non-ENOENT filesystem error occurs` — replaces a session file with a directory (EISDIR) and verifies the import throws without tombstoning.

## P2-2: Defer tombstones until after scan and import succeed

Tombstone session IDs are now collected during the scan and applied only after `importEntries()` completes. If a later scan step fails (e.g. missing transcript marker) or the import throws, no tombstones are committed — the malformed session will be retried on the next launch instead of being permanently suppressed.

**Test added**: `does not tombstone a malformed header when a later scan step fails` — creates a malformed header + an orphan transcript-marker directory with no SQLite metadata, verifies the scan failure prevents tombstoning, then removes the orphan and verifies a re-run succeeds and tombstones correctly.

All 6 tests pass.

**Comment by sunheyi6** [comment]
Updated to address the second round of P2 comments — thank you for catching this.

## Skip without tombstoning (both P2 comments)

`destination.remove()` uses the same tombstone as user deletion, which permanently suppresses a session even if its header is later repaired. For malformed legacy content, the scan now simply skips the session (`continue`) without calling `remove()`. This means:

- A malformed session is retried on the next launch (harmless — it will be skipped again until repaired).
- If the header is repaired, the next import picks it up normally.
- No permanent deletion state is written for corrupt data.
- The scan is free of writes, so there is no partial migration to roll back.

## Test updates

- **'skips a malformed header without tombstoning it while importing valid sessions'**: verifies skip-without-tombstone, then repairs the header and verifies it is imported on the next run.
- **'skips a malformed header without tombstoning when a later scan step fails'**: verifies no tombstone is written when a later scan step fails, and re-running after removing the orphan imports the valid session while skipping (not tombstoning) the malformed one.

All 6 tests pass.

**Comment by Astro-Han** [comment]
Unit test is falling, Would need to fix it before merging.

**Comment by sunheyi6** [comment]
> Unit test is falling, Would need to fix it before merging.

done

File: packages/storage/src/session-metadata-transfer.ts [hunk 2]
```diff
@@ -76,10 +89,11 @@ export async function readLegacySessionMetadataEntry(
   sourcePath: string,
   sessionId: string,
 ): Promise<SessionMetadataImportEntry | null> {
+  // File I/O errors (EACCES, EIO, etc.) propagate without wrapping so they
+  // are not mistaken for corrupt data by isMalformedLegacySessionHeader().
+  const headerLine = await readFirstJsonlRecord(sourcePath);
```
Comment (Astro-Han) [inline packages/storage/src/session-metadata-transfer.ts:99]: [P1] Keep truncated headers on the malformed-content path

Separating file I/O failures from parse and decode failures fixes the tombstoning problem. One corrupt-content case still bypasses the new classifier: `readFirstJsonlRecord()` throws `Cannot read legacy session header` for an empty file, a file that reaches EOF without `\n`, or a first line over 1 MiB. Because this await is outside the `try` block, that error rejects `SqliteSessionStore.ready`; every `list()` and `readHeader()` call continues to fail, which is the startup failure this PR is intended to avoid.

I reproduced this on the current head with an empty `session.jsonl`: `createSessionStore(root).list()` throws the error above. Please distinguish framing or content errors from open, read, and close I/O errors, skip only the former, and add empty and truncated header coverage.

**Comment by Astro-Han** [comment]
The current schema-error and repair tests cover one branch. To close this out, please make the additional cases assert behavior at the public `SessionStore` boundary rather than only checking that some promise rejects.

I would cover these cases:

- Empty `session.jsonl`: `store.list()` resolves, valid sessions remain visible, the corrupt session is absent, and no tombstone is written.
- Truncated JSON without a newline: the same catalog-level assertions as the empty-file case.
- A first record over the 1 MiB limit: the catalog remains usable and the session is skipped without a tombstone.
- A real `turn_state` first record, matching the failure described in the PR: valid sessions remain available through `createSessionStore()`.
- A file-system failure such as `EISDIR`: assert the actual error code or class, not only `assert.rejects()`, and verify that neither metadata nor a tombstone was written.

For each corrupt-content case, repair the file, reopen the store, and assert that the session is imported. That locks in the user-visible contract and avoids testing only an internal error-message shape.

Also we got to fix typecheck.

**Comment by sunheyi6** [comment]
Addressed the latest review feedback:

## Public SessionStore boundary tests (already on head, reformatted)

Assert behavior through `createSessionStore()` / `store.list()` rather than internal promise rejection shapes:

- **Empty `session.jsonl`**: list resolves, valid session visible, corrupt absent, no tombstone; repair → reimport works
- **Truncated JSON without newline**: same catalog-level contract + repair path
- **First record over 1 MiB**: same + repair path
- **`turn_state` first record** (the original bug): valid sessions remain available + repair path
- **EISDIR**: asserts `error.code === 'EISDIR'`, and verifies neither metadata nor tombstone was written

## Typecheck / format

Biome `format:check` was failing on the boundary-test helpers. Pushed a format-only fix (`ad8a7d68`).

CI is green: typecheck / test / e2e all pass.
